src/uri/UriValidator.cpp

```diff
@@ -0,0 +1,127 @@
+#include "UResource.h"
+#include "LongUriSerializer.h"
+#include <string.h>
+
+using namespace uprotocol::uri;
+using namespace uprotocol::utransport;
+
+/**
+ * isValidUri a {@link uri} to ensure that it has at least a name for the
+ * uEntity.
+ *
+ * @param uri {@link uri} to validate.
+ * @return Returns bool containing a true or a false
+ */
+bool UriValidator::valid_uri(const std::string& uri) {
+    auto uri_view = LongUriSerializer::deserialize(uri);
+    return !uri_view.isEmpty();   
+}
+
+/**
+ * Validate a {@link UUri} to ensure that it has at least a name for the uEntity.
+ *
+ * @param uri {@link UUri} to validate.
+ * 
+ * @return Returns Status containing a success or a failure with the error message.
+ */
+UStatus UriValidator::validate(const UUri& uri) {
+  if (uri.isEmpty()) {
+    return UStatus::INVALID_ARGUMENT;
+  }
+  if (!uri.getUAuthority().isEmpty() && !uri.getUAuthority().isRemote()) {
+    return UStatus::INVALID_ARGUMENT;
+  }
+  if (uri.getUEntity().getName().empty()) {
+    return UStatus::INVALID_ARGUMENT;
+  }
+
+  return UStatus::OK;
+}
+
+/**
+ * Validate a {@link UUri} that is meant to be used as an RPC method URI. Used in Request sink values and Response source values.
+ *
+ * @param uri {@link UUri} to validate.
+ * @return Returns Status containing a success or a failure with the error message.
+ */
+UStatus UriValidator::validateRpcMethod(const UUri& uri) {
+    UStatus status = UriValidator::validate(uri);
+    if (status != UStatus::OK) {
+        return status;
+    }
+    if (!UriValidator::isRpcMethod(uri)) {
+        return UStatus::INVALID_ARGUMENT;
+    }
+    
+    return UStatus::OK;
+}
+
+/**
+ * Validate a {@link UUri} that is meant to be used as an RPC method URI. Used in Request sink values and Response source values.
+ *
+ * @param uri {@link UUri} to validate.
+ * @return Returns Status containing a success or a failure with the error message.
+ */
+UStatus UriValidator::validateRpcResponse(const UUri& uri) {
+    UStatus status = UriValidator::validate(uri);
+    if (status != UStatus::OK) {
+        return status;
+    }
+    if (!UriValidator::isRpcResponse(uri)) {
+        return UStatus::INVALID_ARGUMENT;
+    }
+    
+    return UStatus::OK;
+}
+
+/**
+ * Returns true if URI is of type RPC.
+ *
+ * @param uri {@link UUri} to check if it is of type RPC method
+ * @return Returns true if URI is of type RPC.
+ */
+
+bool UriValidator::isRpcMethod(const UUri& uri) {
+    return (!uri.isEmpty() && 
+     uri.getUResource().isRPCMethod() && 
+     !uri.getUResource().getInstance().empty() ||
+     (uri.getUResource().getId().has_value()  && uri.getUResource().getId() != 0));
+}
+
+/**
+ * Returns true if URI is of type RPC response.
+ *
+ * @param uri {@link UUri} to check response
+ * @return Returns true if URI is of type RPC response.
+ */
+bool UriValidator::isRpcResponse(const UUri& uri) {
+    return (UriValidator::isRpcMethod(uri) && 
+    !uri.getUResource().isEmpty() &&
```

Comment:
Actually if isRpcMethod is true then do we need to check isEmpty() (contradicting myself from the above comment)?